## 协程与任务

### 定义协程(coroutine)

#### 单个协程

通过 async/await 语法来声明协程是编写 asyncio 应用的推荐方式。

```python
"""定义并运行最简单的单个协程函数"""
import asyncio
async def main():
    print('hello')
    await asyncio.sleep(1)
    print('world')
asyncio.run(main())
```

#### 协程函数顺序运行

一个函数中运行多个协程时，执行为顺序等待效果。

```python
# 最终执行时间为两个协程执行时间总和
import asyncio
import time

async def display_time(num:int, timeout:int):
    print(f'hello world {num}')
    await asyncio.sleep(timeout)

async def main():
    print(f"started at {time.strftime('%X')}")
    await display_time(1,1)
    await display_time(2,2)
    print(f"end at {time.strftime('%X')}")

asyncio.run(main())
```

#### 协程函数并行运行

[`asyncio.create_task()`](https://docs.python.org/zh-cn/3/library/asyncio-task.html#asyncio.create_task "asyncio.create_task")函数用来并发运行作为 asyncio任务的多个协程。

```python
# 最终执行时间为两个协程执行时间最小公约数
import asyncio
import time

async def display_time(num:int, timeout:int):
    print(f'hello world {num}')
    await asyncio.sleep(timeout)

async def main():
    task1 = asyncio.create_task(display_time(1,1))
    task2 = asyncio.create_task(display_time(2,2))
    print(f"started at {time.strftime('%X')}")
    await task1
    await task2
    print(f"end at {time.strftime('%X')}")

asyncio.run(main())
```

### 可等待对象

如果一个对象可以在 [`await`](https://docs.python.org/zh-cn/3/reference/expressions.html#await) 语句中使用，那么它就是**可等待**对象。许多 asyncio API 都被设计为接受可等待对象。

*可等待* 对象有三种主要类型:  **协程** , **任务** 和  **Future** .

#### 协程

在异步编程中，协程函数是非阻塞的，它们可以在等待某些操作完成的同时继续执行其他任务。当一个协程函数调用另一个协程函数时，如果不等待它的返回结果，那么调用者将继续执行后续的代码，而不会等待被调用的协程函数完成。

如果您需要获取另一个协程函数的返回结果，可以使用 `await`关键字来等待该协程函数的执行完成，并获取其返回值。通过在调用协程函数的地方使用 `await`关键字，可以确保调用者在获取返回结果之前会等待被调用的协程函数执行完毕。

总结：若不使用await调用协程函数，则协程函数无法执行完毕。

#### 任务

*任务*被用来“并行的”调度协程

当一个协程通过 [`asyncio.create_task()`](https://docs.python.org/zh-cn/3/library/asyncio-task.html#asyncio.create_task "asyncio.create_task") 等函数被封装为一个任务，该协程会被自动调度执行:

```python
# create_task自动调度执行协程函数
import asyncio
async def nested():
    return 42

async def main():
    task = asyncio.create_task(nested())
    await task
    print(task.result())

asyncio.run(main())
```

自动执行的含义：

```python
# creat_task将协程函数包装成任务后会立刻执行协程函数，若时间足够则可以执行完毕
import asyncio
async def my_coroutine():
    print("开始执行协程任务")
    await asyncio.sleep(1)
    print("协程结束")
async def main():
    print("开始主程序")
    task = asyncio.create_task(my_coroutine())
    await asyncio.sleep(2)
    print("继续执行主程序")
    await task
    print("协程任务执行完成")

asyncio.run(main())
```

#### Futures

### 等待

**asyncio.**sleep**(***delay**)总是会挂起当前任务，以允许其他任务运行。***

在异步编程中，协程函数之间的切换是由事件循环（Event Loop）控制的。当一个协程函数处于运行状态时，它会占用事件循环的执行权，直到该协程函数主动释放执行权或者等待某个异步操作完成。

使用 `asyncio.sleep(0)`可以将执行权释放给其他等待的协程函数。尽管它的参数是0，看起来似乎是立即完成的，但实际上它会让出一小段时间给其他协程函数执行。这样可以提高并发性能，防止某个协程函数长时间占用事件循环而导致其他协程函数无法得到执行的情况。

### 并发运行任务

asyncio.gather(****aws***)  *并发* 运行 *aws* 序列中的可等待对象(直接协程函数，或task封装的任务)。用于同时运行多个协程，并等待它们全部完成。会返回一个由所有协程返回值组成的列表。

并行运行多个协程函数，效果类似于将每个协程函数利用create_task封装成任务并等待：

```python
import asyncio
import time

async def display_time(num:int, timeout:int):
    print(f'hello world {num}')
    await asyncio.sleep(timeout)
    return num

async def main():
    print(f"started at {time.strftime('%X')}")
    ls = await asyncio.gather(display_time(1, 1), display_time(2, 2))
    print(f"end at {time.strftime('%X')}")
    print(ls)

asyncio.run(main())
```

### 协程保护

asyncio.shield()可以使得保护协程不被取消。

### 超时保护

当使用 `asyncio.timeout()`包装一个协程时，如果该协程在指定的超时时间内没有完成，将会引发 `asyncio.TimeoutError`异常。这可以用于设置协程的最大执行时间，以避免协程长时间阻塞或无限等待的情况。

固定格式(两种)：

```python
async def main():
    async with asyncio.timeout(10):
        await long_running_task()
```

```python
await asyncio.wait_for(long_running_task(), timeout=10)
```

超出

```python
# 当display_time函数运行时长超出asyncio.timeout()时报TimeoutError
import asyncio
async def display_time(timeout):
    await asyncio.sleep(timeout)
    print('OK')

async def main():
    try:
        async with asyncio.timeout(2):
            await display_time(3)
	# 另一种格式
	# await asyncio.wait_for(display_time(3), timeout=2)
    except TimeoutError:
        print("The long operation timed out, but we've handled it.")

    print("This statement will run regardless.")
asyncio.run(main())
```

### 同步运行阻塞函数(time.sleep)

`asyncio.to_thread()`(3.9或更高版本)是 `asyncio`模块中的一个函数，用于在异步上下文中执行阻塞的同步函数。它允许将同步函数转换为协程，并在单独的线程中执行，以避免阻塞事件循环。

blocking_io与display_time并发执行：

```python
# 将同步阻塞函数转换成可等待对象，使得其可以并发执行
import asyncio, time
def blocking_io():
    print(f"start blocking_io at {time.strftime('%X')}")
    # Note that time.sleep() can be replaced with any blocking
    # IO-bound operation, such as file operations.
    time.sleep(1)
    print(f"blocking_io complete at {time.strftime('%X')}")

async def display_time(timeout):
    await asyncio.sleep(timeout)
    print('OK')

async def main():
    print(f"started main at {time.strftime('%X')}")

    await asyncio.gather(
        asyncio.to_thread(blocking_io),
        display_time(2))

    print(f"finished main at {time.strftime('%X')}")

asyncio.run(main())
```

### 跨线程调度协程函数

`asyncio.run_coroutine_threadsafe()`函数，我们可以在不同的线程中安全地运行协程，实现异步操作和线程间的协作。它特别适用于需要在主线程之外执行协程的情况。

此函数应该从另一个 OS 线程中调用，而非事件循环运行所在线程。

## 事件循环


#### done()

判断任务是否完成

#### result

协程函数的返回值



#### add_done_callback

添加一个回调，将在 Task 对象 *完成* 时被运行

```python
import asyncio

async def my_coroutine():
    raise ValueError("An error occurred")

def handle_exception(task):
    try:
        task.result()  # 获取协程函数的返回结果，会引发异常
    except ValueError as e:
        print(f"Caught exception: {str(e)}")

async def main():
    task = asyncio.create_task(my_coroutine())
    task.add_done_callback(handle_exception)

    await asyncio.sleep(1)

asyncio.run(main())
```


#### get_coro()


```python
import asyncio
Keep_Running = True

async def display_time():
    global Keep_Running
    try:
        while(Keep_Running):
            print('Coroutine keep running!')
            await asyncio.sleep(1)
    except asyncio.CancelledError:
        print('Coroutine stop!')
        Keep_Running = False
    finally:
        # 无论是否报错都会执行(跳出try中的循环后)
        print('Coroutine stopped!')
        Keep_Running = False

async def main(cor):
    task_cor = asyncio.create_task(cor)
    # get_coro()返回的是协程函数
    print(f'The current coroutine is {cor_test == task_cor.get_coro()}')
    await asyncio.sleep(3)
    # 取消task_cor，协程函数报错：asyncio.CancelledError
    task_cor.cancel()
    await task_cor
  
cor_test = display_time()
asyncio.run(main(cor_test)
```
